File: 234-palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
    

        empty_list = []

        curr = head

        while curr:
            empty_list.append(curr.val)
            curr = curr.next

        pal_list = empty_list[::-1]

        return empty_list == pal_list

File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {

        ListNode slow = head;

        ListNode fast = head;

        while(fast != null && fast.next != null){

            slow = slow.next;

            fast = fast.next.next;
        }

        // Reverse Second half

        ListNode Second_half = reverseLL(slow);

        ListNode First_half = head;

        while(Second_half != null){

            if(First_half.val != Second_half.val){
                return false;
            }

            First_half = First_half.next;

            Second_half = Second_half.next;
        }

        return true;
        
    }

    public  ListNode reverseLL(ListNode head){

        ListNode prev = null;

        ListNode curr = head;


        while(curr != null){

            ListNode next = curr.next;

            curr.next = prev;

            prev = curr;

            curr = next;

        }

        return prev;
    }
}